Extend ScriptGenerator for verification scenarios

MattWindsor91 opened the pull request:
This PR aims to pull up some functionality relating to generating scripts for `OracleScenario`s and `ScenarioWithQueries`es, as well as other things that we've found necessary for supporting verification related script generation in Azuki natively.

The main points are:

- Added a new subclass `VerificationCapableScriptGenerator` that captures the base functionality of a script generator that can handle these kinds of scenarios.  (I decided not to add it to `ScriptGenerator` because it would pull in several factories that aren't needed by most script generators.)
- Added a mechanism to slipstream in preprocessing for `then` block checks.  I don't really like how this is implemented in the PR as is, but I'm unsure of a better solution (comments welcome).  Ideally I'd have liked to pull in a new `DeclarationState` style setup, but this would require making a new subclass just for `ScriptGenerator`s with complex then blocks.
- Moved some things from the graph scenario to the main library, because they're used downstream.
- Added new interfaces for the various things that can be script-generated in a verification setup, such as queries and action generators.
- In some cases, made things that were previously private part of the API, or opened them where they were closed.  This is because downstream uses them directly, usually.

Questions for review:

- Can we avoid opening and publicising as much internal workings of ScriptGenerator as I've done here?  Maybe we need a separate lower level 'formatting' class?
- Can we do better than `ThenBuilder`?
- Can we name these things better?
- Do we need examples and tests for this?

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationActionGenerator.kt:
```diff
@@ -0,0 +1,7 @@
+
+import com.anaplan.engineering.azuki.core.system.ActionGenerator
+
+interface ScriptGenerationActionGenerator: ActionGenerator {
+    fun getActionGeneratorString(): String
+}
```
🛠️  oh this is wrong.  BRB

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -15,7 +14,7 @@ abstract class ScriptGenerator<
     >(
     private val actionFactory: AF,
     private val checkFactory: CF,
-    private val declarationStateFactory: DeclarationStateFactory<S>
```
❓ H'm.  I can't remember if having these is house style or not.

sifraser replied:
I'm not sure what _these_ is ;)

MattWindsor91 replied:
Nor am I, to be honest, looking back!  😱 

I assume if nothing jumped out here in review it's probably fine though...

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationAction.kt:
```diff
@@ -1,9 +1,27 @@
+class ScriptGenerationParallelAction(actions: List<List<ScriptGenerationAction>>) :
+    ParallelAction<ScriptGenerationAction>(actions), ScriptGenerationAction {
 
+    override fun getActionScript() = """
+        parallel(${
+        runActionsSequentially { it.getActionScript() }.joinToString(", ") {
+            """
+            {
+                ${it.joinToString("\n")}
+            }
+            """
+        }
+    })
+    """
+}
+
+fun Action.toScriptGenAction() =
+    this as? ScriptGenerationAction ?: throw IllegalArgumentException("Incompatible action: $this")
```
These have been hoisted out of the graph example because I know of at least one place where I'll be using them elsewhere.  Whether this is the correct place in Azuki to put them, I'm unsure.

sifraser replied:
I agree that it feels like the 'standard' (only reasonable) way to scriptify the parallel blocks so good to share and this place is no worse than anywhere else :+1:

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -26,60 +25,55 @@ abstract class ScriptGenerator<
+        answer.createValidationChecks(checkFactory)
+
+    open fun getChecksFromAnswer(answer: Answer<*, CF>): List<Check> =
+        answer.createChecks(checkFactory)
```
❓ Do these need to be `open`?  What was the previous rationale for `getChecks` being `open`?  (Are they even necessary?)

❓ Is it ok to weaken the guarantee on the rhs of `getChecks`?  I've become suddenly quite conscious that this breaks API.  At the same time, I suspect downstream only overrides it and doesn't use it.

sifraser replied:
There's been a lot of evolution here and tbf the script gen is usually a 2nd class citizen, so it's good to get a fresh pair of :eyes: and these are likely just artefacts of change

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -26,60 +25,55 @@ abstract class ScriptGenerator<
+        })
+
+    fun generateThenScriptFromChecks(checks: List<Check>): String {
+        val body = thenBuilder().apply { addChecks(checks) }.build()
```
This is my current attempt to slipstream in the ability to group/rearrange/rewrite checks when they go into a then script, which we need downstream.  I'm not massively convinced it's the right way to do this (maybe it should just be a single lambda, or maybe it needs to be closer to `DeclarationState`; thoughts?), but it has the benefit of working enough to be a conversation starter.

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -26,60 +25,55 @@ abstract class ScriptGenerator<
-            else -> throw IllegalArgumentException("Unsupported scenario $scenario")
+        return if (scenario is VerifiableScenario<*, *>) {
+            @Suppress("UNCHECKED_CAST") val then = generateThenScript(scenario as VerifiableScenario<AF, CF>)
+            generateVerifiableScenarioScript(given, whenever, then)
+        } else {
+            generateNonVerifiableScenario(given, whenever, scenario)
```
This is me trying to avoid having to call `super.generateUnformattedScenarioScript` in downstream classes.

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ThenBuilder.kt:
```diff
@@ -0,0 +1,42 @@
+    override fun addCheck(check: Check) {
+        when (check) {
+            is ScriptGenerationCheck -> checks.add(check)
+            is UnsupportedCheck -> Log.debug("Tried to emit unsupported check, ignored")
```
❓ Should `UnsupportedCheck`s be ignored here?

It's important that we have the ability to allow `UnsupportedCheck` to be ignored by _implementations_ of `ThenBuilder`, to avoid needing to all-or-nothing implement check scriptification... but it's not necessarily the case that we need this support by default.

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ThenBuilder.kt:
```diff
@@ -0,0 +1,42 @@
+    fun addCheck(check: Check)
+
+    fun build(): List<ScriptGenerationCheck>
+}
```
As mentioned previously, this could very well just be implemented as `fun buildThen(raw: List<Check>): List<ScriptGenerationCheck>`.  This version makes it easier to implement state machines, but the way it's being used in this PR arguably makes that a bit redundant.

MattWindsor91 commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -128,31 +128,125 @@ abstract class ScriptGenerator<
     }
+}
 
-    fun generateOracleScenarioScript(given: String, whenever: String, generate: String, verify: String): String {
-        return """
```
❓ This was in the public API, should I reimplement it?

MattWindsor91 commented:
I guess in an ideal world my design for checks would look very similar to declaration states:

- add `ScriptGenerationCheckState` as the interface of something that can be fed checks and returns `ScriptGenerationChecks`
- make a distinction between `ScriptGenerationCheck` (has a stringification) and a (needs a better name) `GenerableCheck<S: ScriptGenerationCheckState>`
- push the logic of how to add oneself to `S` into the `GenerableCheck` rather than needing to have a giant pile of logic in the implementation of `ThenBuilder`

Backforming this into Azuki would be awkward, though - it'd introduce a new type parameter into lots of places, and it's hard to know how far to push it, because different adapters will need to do different amounts of polishing up of the checks.  I'll have a think.

**EDIT**: I've added this into the most recent commit, but it needs some work, I think.

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationCheckState.kt:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * Pass through a check that doesn't need any special handling in the check state.
+     */
+    fun addCheck(check: ScriptGenerationCheck)
```
:pick: not sure I understand the comment in relation to the function

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationCheckState.kt:
```diff
@@ -0,0 +1,68 @@
+                is UnsupportedCheck -> checkState.unsupportedCheck()
+                else -> throw IllegalArgumentException("invalid check for script generation: $it")
+            }
+        }
```
:bulb: I don't think the inter-type order would matter here so you could prob use 3/4 variants of:
```
checks.filterIsInstance<GenerableCheck<S>>().forEach { it.generated(checkState) } 
```
to avoid the unchecked cast

MattWindsor91 replied:
I've tried this but I think I've probably made it a bit uglier in the process - I'm unsure how best to catch the 'none of the above' check at this point.

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationCheckState.kt:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Check state for script generation adapters that don't need generable checks.
+ */
+class SimpleScriptGenerationCheckState(private val checks: MutableList<ScriptGenerationCheck> = mutableListOf()):
```
:pick: are you actually expecting this to be pre-initialised? I'd have thought it more likely that this should belong in the class body

MattWindsor91 replied:
Ah yeah oops.  I keep forgetting that putting things up there puts them in the constructor - this is what happens when you come from languages that either have syntactically heavy constructors or no constructors at all 😅

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -1,23 +1,23 @@
 import com.anaplan.engineering.azuki.script.formatter.ScenarioFormatter
 
-abstract class ScriptGenerator<
+abstract class ScriptGeneratorWithCheckState<
```
:pick: Unless this is an _additional_ class, I don't see the value in the name change

sifraser replied:
Found  the other script generator and the naming feels the wrong way around to me .. i.e. it's the 'adapter' that should have the exceptional name

MattWindsor91 replied:
Knocked the two into one in update.  Breaks API but this should be ok.

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerationCheckState.kt:
```diff
@@ -0,0 +1,68 @@
+
+/**
+ * State used to build a 'then' block out of a stream of low-level generator or generable checks.
+ */
```
:pick: I feel that this description is a little too use-case specific.. surely it's just a state that enables the composition of composable checks for 'efficiency', where efficiency is somewhat in the eye of the beholder.

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -26,60 +26,53 @@ abstract class ScriptGenerator<
-        scenario.checks(checkFactory).map { it as ScriptGenerationCheck }
+    protected open fun generateNonVerifiableScenario(
+        given: String, whenever: String, scenario: BuildableScenario<AF>
+    ): String = throw IllegalArgumentException("Unsupported scenario $scenario")
```
```suggestion
    ): String = throw UnsupportedOperationException("Unsupported scenario type ${scenario::class}")
```

sifraser commented on azuki-script-generation/src/main/kotlin/com/anaplan/engineering/azuki/script/generation/ScriptGenerator.kt:
```diff
@@ -26,60 +26,53 @@ abstract class ScriptGenerator<
+        answer.createChecks(checkFactory)
 
-    fun generateThenScript(scenario: VerifiableScenario<AF, CF>) =
+    private fun generateThenScript(scenario: VerifiableScenario<AF, CF>) =
```
:question: not sure what the driver is for making this private.. I'd have assumed it was called from outside

MattWindsor91 replied:
Woops, I'm not sure either.  It's not in keeping with the other such functions.  Reverted.

sifraser commented:
> Can we avoid opening and publicising as much internal workings of ScriptGenerator as I've done here? Maybe we need a separate lower level 'formatting' class?

It's probably a question of determining how much is common to all users and how much is just a utility .. as per previous comment doing the tictactoe would provide another usecase to assist

> Can we do better than ThenBuilder?

I think this has gone

> Can we name these things better?

Probably :)

> Do we need examples and tests for this?

Certainly think adding into tictactoe would be helpful...

Testing at this high level of abstraction is hard .. I should imagine you find that you'd need an example usecase to do the testing..

MattWindsor91 commented:
Gonna reroll this as I've tried to address the reviews but also added a substantial wad of extra code for exercising this against tictactoe.  Some of that is VDM and I'm no good at VDM so it likely needs a bit of work.

Also:

- Renamed `GenerableCheck` to `ComposableCheck` to try to lean into this idea that a check-state _composes_ checks together
- Added 'board has token N at position' and 'board has space at position' checks to tic-tac-toe.  The existing 'board has state' is _not_ lowered into these right now, since it's more efficient to check on some implementations (see below).
- The example for use of a check-state sees whether all 9 places on the board have either a token or a space specified for them, and combines them into a state check if so.  I have a (somewhat nastily written) unit test to make sure this happens!  This is somewhat contrived given the above but I could always force a lowering of has-state to has-token+has-space to _really_ exercise the flow.

In doing this I wonder if it's worth at some point pulling the check-state stuff out and making it possible to use it in implementations as well.  The VDM specifications I've written for board-has-token and board-has-space are very wasteful, and one could imagine that in some uses of Azuki there's legitimate merit to composing some checks (or optimising them!) before presenting them to an oracle.  I won't do that in this PR though as it needs some thoughtful design.
